pkg/ephemeral/io/tuple_streamer.go

```diff
@@ -0,0 +1,242 @@
+	gameUUID, err := uuid.Parse(gameID)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing gameID: %v", err)
+	}
+	return &CastorTupleStreamer{
+		logger:        l,
+		pipeWriter:    pipeWriter,
+		tupleType:     t,
+		stockSize:     conf.TupleStock,
+		castorClient:  conf.CastorClient,
+		baseRequestID: uuid.NewMD5(gameUUID, []byte(t)),
+		streamData:    headerData,
+	}, nil
+}
+
+// CastorTupleStreamer provides tuples to the SPDZ execution for the given type and configuration.
+type CastorTupleStreamer struct {
+	logger        *zap.SugaredLogger
+	pipeWriter    PipeWriter
+	tupleType     castor.TupleType
+	stockSize     int32
+	castorClient  castor.AbstractClient
+	baseRequestID uuid.UUID
+	requestCycle  int
+	streamData    []byte
+}
+
+// StreamTuples repeatedly downloads a given type of tuples from castor and streams it to the according file as required by MP-SPDZ
+func (ts *CastorTupleStreamer) StartStreamTuples(terminate chan struct{}, errCh chan error, wg *sync.WaitGroup) {
+	defer func() {
+		ts.logger.Debugw(fmt.Sprintf("Done streaming tuples, discarding %d bytes.", len(ts.streamData)), "TupleType", ts.tupleType)
+		ts.pipeWriter.Close()
+		wg.Done()
+	}()
+	for {
+		select {
+		case <-terminate:
+			return
+		default:
+			err := ts.writeDataToPipe()
+			if err != nil {
+				errCh <- err
+				return
+			}
+		}
+	}
+}
+
+func (ts *CastorTupleStreamer) writeDataToPipe() error {
+	if ts.streamData == nil || len(ts.streamData) == 0 {
+		requestID := uuid.NewMD5(ts.baseRequestID, []byte(strconv.Itoa(ts.requestCycle))).String()
+		ts.requestCycle++
+		tupleList, err := ts.castorClient.GetTuples(ts.stockSize, ts.tupleType, requestID)
+		if err != nil {
+			return err
+		}
+		ts.streamData, err = ts.tupleListToByteArray(tupleList)
+	}
+	c, err := ts.pipeWriter.Write(ts.streamData)
+	if err != nil {
+		ts.logger.Errorw(err.Error(), "TupleType", ts.tupleType)
+	}
+	ts.streamData = ts.streamData[c:]
+	ts.logger.Debug(fmt.Sprintf("Wrote %d %s to file %s", c, ts.tupleType, ts.pipeWriter.(*TuplePipeWriter).tupleFile.Name()))
+	return nil
+}
+
+func (ts *CastorTupleStreamer) tupleListToByteArray(tl castor.TupleList) ([]byte, error) {
+	var result []byte
+
+	for _, tuple := range tl.Tuples {
+		for _, share := range tuple.Shares {
+			decodeString, err := base64.StdEncoding.DecodeString(share.Value)
+			if err != nil {
+				return []byte{}, err
+			}
+			result = append(result, decodeString...)
+
+			decodeString, err = base64.StdEncoding.DecodeString(share.Mac)
+			if err != nil {
+				return []byte{}, err
+			}
+			result = append(result, decodeString...)
+		}
+	}
+
+	return result, nil
+}
+
+func protocolDescriptorFor(tt castor.TupleType) string {
+	var typeString string = fmt.Sprint(tt)
+	return fmt.Sprint("SPDZ ", strings.ToLower(typeString[strings.LastIndex(typeString, "_")+1:]))
+}
+
+func generateHeader(tt castor.TupleType, prime *big.Int) []byte {
```

Comment:
This currently only generates GFP Headers, since the prime number is not relevant for GF2N, we'd need to check what the header contains (maybe the "securityParameter"/Bitlength?)

Reply:
Gf2n header contains
- Length to follow (little-endian 8-byte number)
- Protocol descriptor
- Domain descriptor

where Protocol descriptor is `SPDZ gf2n_` and domain descriptor loots as follow:
  - Storage size in bytes (little-endian 8-byte number). Default is 16.
  - :math:`n` (little-endian 4-byte number)

Reply:
gf2n file headers created according to definition